lms/djangoapps/experiments/flags.py

```diff
@@ -0,0 +1,133 @@
+        if cache_response.is_found:
+            return cache_response.value
+
+        # Check if the main flag is even enabled for this user and course.
+        if not self._is_enabled(course_key):  # grabs user from the current request, if any
+            return self._cache_bucket(experiment_name, 0)
+
+        # Check if the enrollment should even be considered (if it started before the experiment wants, we ignore)
+        if course_key and self.experiment_id is not None:
+            start_val = ExperimentKeyValue.objects.filter(experiment_id=self.experiment_id, key='enrollment_start')
+            if start_val:
+                try:
+                    start_date = dateutil.parser.parse(start_val.first().value).replace(tzinfo=pytz.UTC)
+                except ValueError:
+                    log.exception('Could not parse enrollment start date for experiment %d', self.experiment_id)
+                    return self._cache_bucket(experiment_name, 0)
+                enrollment = CourseEnrollment.get_enrollment(request.user, course_key)
+                # Only bail if they have an enrollment and it's old -- if they don't have an enrollment, we want to do
+                # normal bucketing -- consider the case where the experiment has bits that show before you enroll. We
+                # want to keep your bucketing stable before and after you do enroll.
+                if enrollment and enrollment.created < start_date:
+                    return self._cache_bucket(experiment_name, 0)
+
+        bucket = stable_bucketing_hash_group(experiment_name, self.num_buckets, request.user.username)
+
+        # Now check if the user is forced into a particular bucket, using our subordinate bucket flags
+        for i, bucket_flag in enumerate(self.bucket_flags):
+            if bucket_flag.is_enabled(course_key):
+                bucket = i
+                break
+
+        session_key = 'tracked.{}'.format(experiment_name)
+        if track and hasattr(request, 'session') and session_key not in request.session:
+            segment.track(
+                user_id=request.user.id,
+                event_name='edx.bi.experiment.user.bucketed',
+                properties={
+                    'site': request.site.domain,
+                    'app_label': self.waffle_namespace.name,
+                    'experiment': self.flag_name,
+                    'course_id': str(course_key) if course_key else None,
+                    'bucket': bucket,
+                    'is_staff': request.user.is_staff,
+                }
+            )
+
+            # Mark that we've recorded this bucketing, so that we don't do it again this session
+            request.session[session_key] = True
+
+        return self._cache_bucket(experiment_name, bucket)
+
+    def is_enabled(self, course_key=None):
+        return self.get_bucket(course_key) != 0
+
+    def is_enabled_without_course_context(self):
+        return self.get_bucket() != 0
```

Comment:
These are convenience methods, to keep same API as CourseWaffleFlag and for when you only care if the user is not in the control bucket.

This also lets us not have to make any updates to edx-when -- it uses this API for the relative dates experiment, and only cares about the two buckets.